File: frontend/src/utils/pdfGenerator.js
import jsPDF from "jspdf";

export const generatePDF = (data) => {
  const doc = new jsPDF();

  let y = 20;

  doc.setFont("helvetica", "bold");
  doc.setFontSize(20);
  doc.text("AI Resume Screening Report", 20, y);

  y += 15;

  doc.setFontSize(12);

  doc.text(`Name: ${data.name}`, 20, y);
  y += 8;

  doc.text(`Email: ${data.email}`, 20, y);
  y += 8;

  doc.text(`Phone: ${data.phone}`, 20, y);
  y += 8;

  doc.text(`ATS Score: ${data.ats_score}/100`, 20, y);

  y += 12;

  doc.setFont(undefined, "bold");
  doc.text("Skills", 20, y);

  y += 8;

  doc.setFont(undefined, "normal");

  doc.text(data.skills.join(", "), 20, y, {
    maxWidth: 170,
  });

  y += 18;

  doc.setFont(undefined, "bold");
  doc.text("Education", 20, y);

  y += 8;

  doc.setFont(undefined, "normal");

  doc.text(data.education, 20, y, {
    maxWidth: 170,
  });

  y += 18;

  doc.setFont(undefined, "bold");
  doc.text("Experience", 20, y);

  y += 8;

  doc.setFont(undefined, "normal");

  doc.text(data.experience, 20, y, {
    maxWidth: 170,
  });

  y += 25;

  doc.setFont(undefined, "bold");
  doc.text("Projects", 20, y);

  y += 8;

  doc.setFont(undefined, "normal");

  doc.text(data.projects, 20, y, {
    maxWidth: 170,
  });

  y += 25;

  doc.setFont(undefined, "bold");
  doc.text("Professional Summary", 20, y);

  y += 8;

  doc.setFont(undefined, "normal");

  doc.text(data.summary, 20, y, {
    maxWidth: 170,
  });

  y += 25;

  doc.setFont(undefined, "bold");
  doc.text("Strengths", 20, y);

  y += 8;

  doc.setFont(undefined, "normal");

  data.strengths.forEach((item) => {
    doc.text(`• ${item}`, 25, y);
    y += 7;
  });

  y += 5;

  doc.setFont(undefined, "bold");
  doc.text("Weaknesses", 20, y);

  y += 8;

  doc.setFont(undefined, "normal");

  data.weaknesses.forEach((item) => {
    doc.text(`• ${item}`, 25, y);
    y += 7;
  });

  y += 5;

  doc.setFont(undefined, "bold");
  doc.text("Suggestions", 20, y);

  y += 8;

  doc.setFont(undefined, "normal");

  data.suggestions.forEach((item) => {
    doc.text(`• ${item}`, 25, y);
    y += 7;
  });

  doc.save(`${data.name}_Resume_Report.pdf`);
};

File: backend/src/pdf_generator.py
import os

from reportlab.platypus import (
    SimpleDocTemplate,
    Paragraph,
    Spacer,
)

from reportlab.lib.styles import getSampleStyleSheet
from reportlab.lib.enums import TA_CENTER
from reportlab.lib.colors import HexColor


REPORT_FOLDER = "reports"

os.makedirs(REPORT_FOLDER, exist_ok=True)


def generate_pdf(candidate):
    """
    Generates a professional PDF report
    Returns the PDF file path
    """

    filename = (
        candidate["name"]
        .replace(" ", "_")
        .replace("/", "_")
        + "_Report.pdf"
    )

    pdf_path = os.path.join(REPORT_FOLDER, filename)

    doc = SimpleDocTemplate(pdf_path)

    styles = getSampleStyleSheet()

    title = styles["Heading1"]
    title.alignment = TA_CENTER
    title.textColor = HexColor("#0EA5E9")

    heading = styles["Heading2"]

    normal = styles["BodyText"]

    story = []

    # -------------------------------------------------------
    # Title
    # -------------------------------------------------------

    story.append(
        Paragraph(
            "AI Resume Screening Report",
            title
        )
    )

    story.append(Spacer(1, 20))

    # -------------------------------------------------------
    # Basic Details
    # -------------------------------------------------------

    story.append(
        Paragraph(
            "<b>Candidate Name:</b> "
            + candidate.get("name", ""),
            normal,
        )
    )

    story.append(
        Paragraph(
            "<b>Email:</b> "
            + candidate.get("email", ""),
            normal,
        )
    )

    story.append(
        Paragraph(
            "<b>Phone:</b> "
            + candidate.get("phone", ""),
            normal,
        )
    )

    story.append(Spacer(1, 15))

    # -------------------------------------------------------
    # Scores
    # -------------------------------------------------------

    story.append(Paragraph("Scores", heading))

    story.append(
        Paragraph(
            f"<b>ATS Score:</b> {candidate.get('ats_score',0)}%",
            normal,
        )
    )

    story.append(
        Paragraph(
            f"<b>JD Match Score:</b> {candidate.get('jd_match_score',0)}%",
            normal,
        )
    )

    story.append(Spacer(1, 15))

    # -------------------------------------------------------
    # Skills
    # -------------------------------------------------------

    story.append(Paragraph("Skills", heading))

    skills = candidate.get("skills", [])

    for skill in skills:
        story.append(
            Paragraph(f"• {skill}", normal)
        )

    story.append(Spacer(1, 15))

    # -------------------------------------------------------
    # Education
    # -------------------------------------------------------

    story.append(Paragraph("Education", heading))

    story.append(
        Paragraph(
            candidate.get("education", ""),
            normal,
        )
    )

    story.append(Spacer(1, 15))

    # -------------------------------------------------------
    # Experience
    # -------------------------------------------------------

    story.append(Paragraph("Experience", heading))

    story.append(
        Paragraph(
            candidate.get("experience", ""),
            normal,
        )
    )

    story.append(Spacer(1, 15))

    # -------------------------------------------------------
    # Projects
    # -------------------------------------------------------

    story.append(Paragraph("Projects", heading))

    story.append(
        Paragraph(
            candidate.get("projects", ""),
            normal,
        )
    )

    story.append(Spacer(1, 15))

    # -------------------------------------------------------
    # Strengths
    # -------------------------------------------------------

    story.append(Paragraph("Strengths", heading))

    for item in candidate.get("strengths", []):
        story.append(
            Paragraph(f"• {item}", normal)
        )

    story.append(Spacer(1, 15))

    # -------------------------------------------------------
    # Weaknesses
    # -------------------------------------------------------

    story.append(Paragraph("Weaknesses", heading))

    for item in candidate.get("weaknesses", []):
        story.append(
            Paragraph(f"• {item}", normal)
        )

    story.append(Spacer(1, 15))

    # -------------------------------------------------------
    # Suggestions
    # -------------------------------------------------------

    story.append(Paragraph("Suggestions", heading))

    for item in candidate.get("suggestions", []):
        story.append(
            Paragraph(f"• {item}", normal)
        )

    story.append(Spacer(1, 15))

    # -------------------------------------------------------
    # Interview Questions
    # -------------------------------------------------------

    story.append(
        Paragraph(
            "Interview Questions",
            heading,
        )
    )

    for q in candidate.get(
        "interview_questions",
        [],
    ):

        story.append(
            Paragraph(
                f"<b>{q.get('difficulty','')}</b> | "
                f"{q.get('category','')}",
                normal,
            )
        )

        story.append(
            Paragraph(
                q.get("question", ""),
                normal,
            )
        )

        story.append(Spacer(1, 8))

    # -------------------------------------------------------
    # Footer
    # -------------------------------------------------------

    story.append(Spacer(1, 25))

    story.append(
        Paragraph(
            "<b>Generated by AI Resume Screening System</b>",
            normal,
        )
    )

    doc.build(story)

    return pdf_path